js/private/lifecycle/lifecycle-hooks.js

```diff
@@ -0,0 +1,78 @@
+const {
+    mkdirSync,
+    copyFileSync,
+    readdirSync,
+    readFileSync,
+    existsSync,
+    statSync,
+} = require('fs')
+const path = require('path')
+const { runPostinstallHooks } = require('@pnpm/lifecycle')
+
+async function main(argv) {
+    if (argv.length !== 2) {
+        console.error('Usage: node lifecycle-hooks.js [packageDir] [outputDir]')
+        process.exit(1)
+    }
+    const packageDir = argv[0]
+    const outputDir = argv[1]
+
+    copyPackageContents(packageDir, outputDir)
+
+    const packageJson = JSON.parse(
+        readFileSync(path.join(packageDir, 'package.json'))
+    )
+
+    if (
+        packageJson.scripts?.preinstall ||
+        packageJson.scripts?.install ||
+        packageJson.scripts?.postinstall
+    ) {
+        const nodeModulesPath = path.join(outputDir, '..', 'node_modules')
+
+        // If the package we're running postinstall on has no deps, then node_modules
+        // won't exist in the sandbox. The lifecycle functions provided by pnpm require
+        // it to exist, so create an empty folder if needed.
+        if (!existsSync(nodeModulesPath)) {
+            mkdirSync(nodeModulesPath)
+        }
+
+        // Runs preinstall, install, and postinstall hooks
+        await runPostinstallHooks({
+            pkgId: packageJson.name + '/' + packageJson.version,
```

Comment:
no guarantee that package.json has `name` and `version` fields so we should handle it gracefully if they don't. version can be defaulted to `0.0.0` (which I do for first party deps elsewhere). not sure what to do if name doesn't exist. what does `runPostinstallHooks` do with `pkgId`? is it load bearing somehow?

Reply:
This actually isn't a real property. It's in their usage which is just out of date.